From 5c16c69ce9961c393aed212229843f8f0b76dd18 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Fri, 9 Jul 2010 13:34:46 -0600
Subject: [PATCH] Add support for ${user} in refs

Interpret a ${user} string in a ref as the name of the
current user.  This allows the refs namespace to have
generic rules applied to refs involving the user's
name without having to hardcode a rule for each user.

Bug: issue 577
---
 .../google/gerrit/server/project/RefControl.java   |   20 +++++++++++++++-----
 1 files changed, 15 insertions(+), 5 deletions(-)

diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/RefControl.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/RefControl.java
index 68e6412..3bbada4 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/RefControl.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/RefControl.java
@@ -392,7 +392,7 @@ public class RefControl {
   private List<RefRight> filter(Collection<RefRight> all) {
     List<RefRight> mine = new ArrayList<RefRight>(all.size());
     for (RefRight right : all) {
-      if (matches(getRefName(), right.getRefPattern())) {
+      if (matches(getRefName(), right.getRefPattern(), getCurrentUser())) {
         mine.add(right);
       }
     }
@@ -403,13 +403,23 @@ public class RefControl {
     return projectControl.getProjectState();
   }
 
-  public static boolean matches(String refName, String refPattern) {
-    if (refPattern.endsWith("/*")) {
-      String prefix = refPattern.substring(0, refPattern.length() - 1);
+  public static boolean matches(String refName, String refPattern, CurrentUser user) {
+    String userPattern = refPattern;
+    if (user instanceof IdentifiedUser) {
+      String userName = ((IdentifiedUser) user).getUserName();
+      userPattern = refPattern.replaceAll("\\$\\{user\\}", userName);
+    } else {
+      if (refPattern.indexOf("${user}") > -1) {
+        return false;
+      }
+    }
+
+    if (userPattern.endsWith("/*")) {
+      String prefix = userPattern.substring(0, userPattern.length() - 1);
       return refName.startsWith(prefix);
 
     } else {
-      return refName.equals(refPattern);
+      return refName.equals(userPattern);
     }
   }
 }
-- 
1.7.1

